# **NBA Anaysis: Extract, Transform, Load**

## Objectives

* Load data set from Kaggle: 
https://www.kaggle.com/datasets/dbtjdals/nba-mvp-candidates-1980-2022?select=master_table.csv

There are 2 files, one for partial data for 2022, and a larger file with data from 1980-2021, the file that will be used: master_table.csv



## Inputs

* Files used: 
master_table.csv


## Outputs

* File produced:
nba-data-1980plus.csv 

## Additional Comments

* This dataset contains essential information about the wins, losses, typical basketball statistics (points per game, blocks per game, rebounds per game, etc), as well as more advanced statistics (Player Efficiency Rating, Box Plus Minus, Value Over Replacement Player, etc) per each MVP candidate.

* Note: there was another data set with over 37,000 rows, which may have been too large for the scope of this analysis but may be useful to expand and look at trends, eg. how many 3 pointers have been taken per year, which has dramatically increased since the addition of the 3 point line in 1980, with the most 3 pointers taken by a team just 10 years ago being lower than the number of 3 point shots taken by the team that takes the lowest number of 3 point shots these days.
* Note: only require data from 1980 onwards




---

# Section 1

#Import necessary libraries

In [2]:
import pandas as pd

# Load CSV file

In [3]:
df1 = pd.read_csv('../data/inputs/raw/master_table.csv')

# Look at data

In [4]:
df1.head()


Rank               Player  Age   Tm  First  Pts Won  Pts Max  Share   G  \
0    1  Kareem Abdul-Jabbar   32  LAL  147.0    147.0      221  0.665  82   
1    2        Julius Erving   29  PHI   31.5     31.5      221  0.143  78   
2    3        George Gervin   27  SAS   19.0     19.0      221  0.086  78   
3    4           Larry Bird   23  BOS   15.0     15.0      221  0.068  82   
4   5T       Tiny Archibald   31  BOS    2.0      2.0      221  0.009  80   

     MP  ...  turnover_percentage  usage_percentage  offensive_win_shares  \
0  38.3  ...                 15.7              24.1                   9.5   
1  36.1  ...                 13.3              30.3                   7.3   
2  37.6  ...                 10.3              31.7                   9.3   
3  36.0  ...                 14.0              25.3                   5.6   
4  35.8  ...                 19.7              17.0                   5.9   

   defensive_win_shares  win_shares  win_shares_per_48_minutes  \
0                   5.3        14.8                      0.227   
1                   5.2        12.5                      0.213   
2                   1.3        10.6                      0.173   
3                   5.6        11.2                      0.182   
4                   2.9         8.9                      0.148   

   offensive_box_plus_minus  defensive_box_plus_minus  box_plus_minus  \
0                       4.8                       2.4             7.2   
1                       5.7                       1.9             7.6   
2                       5.5                      -1.6             3.9   
3                       3.0                       1.5             4.5   
4                       1.4                      -0.3             1.1   

   value_over_replacement_player  
0                            7.3  
1                            6.8  
2                            4.4  
3                            4.8  
4                            2.3  

[5 rows x 45 columns]

# Look at shape

In [5]:
df1.shape

(679, 45)

# We can see there are 679 rows and 45 columns, 1 for each MVP candidate per year.

# Check data types

In [6]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 679 entries, 0 to 678
Data columns (total 45 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Rank                           679 non-null    object 
 1   Player                         679 non-null    object 
 2   Age                            679 non-null    int64  
 3   Tm                             679 non-null    object 
 4   First                          679 non-null    float64
 5   Pts Won                        679 non-null    float64
 6   Pts Max                        679 non-null    int64  
 7   Share                          679 non-null    float64
 8   G                              679 non-null    int64  
 9   MP                             679 non-null    float64
 10  PTS                            679 non-null    float64
 11  TRB                            679 non-null    float64
 12  AST                            679 non-null    flo

# Check null values

In [7]:
df1.isnull().sum()

Rank                             0
Player                           0
Age                              0
Tm                               0
First                            0
Pts Won                          0
Pts Max                          0
Share                            0
G                                0
MP                               0
PTS                              0
TRB                              0
AST                              0
STL                              0
BLK                              0
FG%                              0
3P%                              0
FT%                              0
WS                               0
WS/48                            0
year                             0
team                             0
W                                0
W/L%                             0
seed                             0
player_efficiency_rating         0
true_shooting_percentage         0
three_point_attempt_rate         0
free_throw_attempt_r

# Check for duplicates

In [14]:
df1.duplicated().sum()

0

# Only data from 1980-2018 onwards is being used, so anything outside this range needs to be removed. The data set begins from 1980 so we need to filter from 2018 and less.
# Filter rows where the year is 2018 or before

In [15]:
filtered_df = df1[df1['year'] <= 2018]


# Check new data frame


In [16]:
filtered_df.head()

Rank               Player  Age   Tm  First  Pts Won  Pts Max  Share   G  \
0    1  Kareem Abdul-Jabbar   32  LAL  147.0    147.0      221  0.665  82   
1    2        Julius Erving   29  PHI   31.5     31.5      221  0.143  78   
2    3        George Gervin   27  SAS   19.0     19.0      221  0.086  78   
3    4           Larry Bird   23  BOS   15.0     15.0      221  0.068  82   
4   5T       Tiny Archibald   31  BOS    2.0      2.0      221  0.009  80   

     MP  ...  turnover_percentage  usage_percentage  offensive_win_shares  \
0  38.3  ...                 15.7              24.1                   9.5   
1  36.1  ...                 13.3              30.3                   7.3   
2  37.6  ...                 10.3              31.7                   9.3   
3  36.0  ...                 14.0              25.3                   5.6   
4  35.8  ...                 19.7              17.0                   5.9   

   defensive_win_shares  win_shares  win_shares_per_48_minutes  \
0                   5.3        14.8                      0.227   
1                   5.2        12.5                      0.213   
2                   1.3        10.6                      0.173   
3                   5.6        11.2                      0.182   
4                   2.9         8.9                      0.148   

   offensive_box_plus_minus  defensive_box_plus_minus  box_plus_minus  \
0                       4.8                       2.4             7.2   
1                       5.7                       1.9             7.6   
2                       5.5                      -1.6             3.9   
3                       3.0                       1.5             4.5   
4                       1.4                      -0.3             1.1   

   value_over_replacement_player  
0                            7.3  
1                            6.8  
2                            4.4  
3                            4.8  
4                            2.3  

[5 rows x 45 columns]

# It can be seen here that names are in one field which may be a potential problem later when linking to other data. Change the filtered data separating out the first name and last name and first initial. 

In [17]:
# Split into Name and Surname
filtered_df[['Name', 'Surname']] = filtered_df['Player'].str.split(' ', n=1, expand=True)

# Create First_Name (first letter of full name + '.')
filtered_df['First_Name'] = filtered_df['Player'].str[0] + '.'

# Reorder columns into a new data frame to see the results
new_df = filtered_df[['Player', 'First_Name', 'Name', 'Surname']]

# View result
print(new_df)

                  Player First_Name      Name       Surname
0    Kareem Abdul-Jabbar         K.    Kareem  Abdul-Jabbar
1          Julius Erving         J.    Julius        Erving
2          George Gervin         G.    George        Gervin
3             Larry Bird         L.     Larry          Bird
4         Tiny Archibald         T.      Tiny     Archibald
..                   ...        ...       ...           ...
637    LaMarcus Aldridge         L.  LaMarcus      Aldridge
638         Jimmy Butler         J.     Jimmy        Butler
639        Stephen Curry         S.   Stephen         Curry
640          Joel Embiid         J.      Joel        Embiid
641       Victor Oladipo         V.    Victor       Oladipo

[642 rows x 4 columns]


C:\Users\ronni\AppData\Local\Temp\ipykernel_39036\3731656790.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df[['Name', 'Surname']] = filtered_df['Player'].str.split(' ', n=1, expand=True)
C:\Users\ronni\AppData\Local\Temp\ipykernel_39036\3731656790.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df[['Name', 'Surname']] = filtered_df['Player'].str.split(' ', n=1, expand=True)
C:\Users\ronni\AppData\Local\Temp\ipykernel_39036\3731656790.py:5: SettingWithCopyWarning: 
A value is try

# Check the giltered data frame to make sure all required data is there

In [18]:
filtered_df.head()

Rank               Player  Age   Tm  First  Pts Won  Pts Max  Share   G  \
0    1  Kareem Abdul-Jabbar   32  LAL  147.0    147.0      221  0.665  82   
1    2        Julius Erving   29  PHI   31.5     31.5      221  0.143  78   
2    3        George Gervin   27  SAS   19.0     19.0      221  0.086  78   
3    4           Larry Bird   23  BOS   15.0     15.0      221  0.068  82   
4   5T       Tiny Archibald   31  BOS    2.0      2.0      221  0.009  80   

     MP  ...  defensive_win_shares  win_shares  win_shares_per_48_minutes  \
0  38.3  ...                   5.3        14.8                      0.227   
1  36.1  ...                   5.2        12.5                      0.213   
2  37.6  ...                   1.3        10.6                      0.173   
3  36.0  ...                   5.6        11.2                      0.182   
4  35.8  ...                   2.9         8.9                      0.148   

   offensive_box_plus_minus  defensive_box_plus_minus  box_plus_minus  \
0                       4.8                       2.4             7.2   
1                       5.7                       1.9             7.6   
2                       5.5                      -1.6             3.9   
3                       3.0                       1.5             4.5   
4                       1.4                      -0.3             1.1   

   value_over_replacement_player    Name       Surname  First_Name  
0                            7.3  Kareem  Abdul-Jabbar          K.  
1                            6.8  Julius        Erving          J.  
2                            4.4  George        Gervin          G.  
3                            4.8   Larry          Bird          L.  
4                            2.3    Tiny     Archibald          T.  

[5 rows x 48 columns]

# Check for null values

In [19]:
filtered_df.isnull().sum()

Rank                             0
Player                           0
Age                              0
Tm                               0
First                            0
Pts Won                          0
Pts Max                          0
Share                            0
G                                0
MP                               0
PTS                              0
TRB                              0
AST                              0
STL                              0
BLK                              0
FG%                              0
3P%                              0
FT%                              0
WS                               0
WS/48                            0
year                             0
team                             0
W                                0
W/L%                             0
seed                             0
player_efficiency_rating         0
true_shooting_percentage         0
three_point_attempt_rate         0
free_throw_attempt_r

# Check for duplicates

In [21]:
filtered_df.duplicated().sum()

0

# Save to a new CSV file

In [22]:
filtered_df.to_csv('../data/outputs/nba-data-1980plus.csv', index=False)

---

# Section 2

Section 2 content

---

NOTE

* You may add as many sections as you want, as long as it supports your project workflow.
* All notebook's cells should be run top-down (you can't create a dynamic wherein a given point you need to go back to a previous cell to execute some task, like go back to a previous cell and refresh a variable content)

---

# Push files to Repo

* In cases where you don't need to push files to Repo, you may replace this section with "Conclusions and Next Steps" and state your conclusions and next steps.

In [ ]:
import os
try:
  # create your folder here
  # os.makedirs(name='')
except Exception as e:
  print(e)
